Remove internal option layout.autosize='initial' (Fixes #537)

n-riesco opened the pull request:
* Moved initial call to `plotAutoSize` into `Plots.supplyDefaults(gd)`.

* Replaced `{ autosize: 'initial' }` with the flag
  `gd._fullLayout._initialAutoSizeIsDone`.

* `{ autosize: false }` the values of width and height undefined in
  `gd.layout` will be autosized only once.

* `{ autosize: true }` only autosizes the values of width and height
  undefined in `gd.layout`.

Fixes #537

etpinard commented on src/plots/plots.js:
```diff
@@ -752,6 +780,85 @@ plots.supplyLayoutGlobalDefaults = function(layoutIn, layoutOut) {
+        newWidth,
+        newHeight;
+
+    if(typeof gd.emit === 'function') gd.emit('plotly_autosize');
```
why do we need this?

etpinard replied:
i.e. that `if(typeof gd.emit === 'function')`

n-riesco replied:
Otherwise I get errors in the console (perhaps some test are mocking gd with something other than a DOM element?). I will come back to you on this tomorrow.

etpinard replied:
Good call. `Plots.supplyDefaults` should work with plain objects.

Could we make this check more robust? I'd vote for using [`Lib.isPlotDiv`](https://github.com/plotly/plotly.js/blob/19a37357341e7e098ce2c20ee98f65965c2c5d66/src/lib/index.js#L380).

etpinard commented on src/plots/plots.js:
```diff
@@ -752,6 +780,85 @@ plots.supplyLayoutGlobalDefaults = function(layoutIn, layoutOut) {
+        // specify size in layout, or take the defaults,
+        // but don't enforce any ratio restrictions
+        var computedStyle;
+        try {
```
why do we need a try-catch ?

n-riesco replied:
same reason as above. I will come back to you with more details on this tomorrow.

n-riesco replied:
For example, [here](https://github.com/plotly/plotly.js/blob/476818c5ff6c3c7301bb897c5af13ade83c848d0/test/jasmine/tests/plot_api_test.js#L71)'s a test that uses a mocked `gd`. Since `gd` isn't a DOM element the call to `window.getComputedStyle(gd)` throws a `TypeError`.

alexcjohnson replied:
I think @chriddyp and possibly others have good reasons for calling `supplyDefaults` outside the DOM so it does seem reasonable to support (and test) this behavior

n-riesco replied:
I've also noticed that `restyle` in master also throws an error [here](https://github.com/plotly/plotly.js/blob/476818c5ff6c3c7301bb897c5af13ade83c848d0/src/plot_api/plot_api.js#L2038) when tests like [this](https://github.com/plotly/plotly.js/blob/476818c5ff6c3c7301bb897c5af13ade83c848d0/test/jasmine/tests/plot_api_test.js#L262) use a mocked `gd`. Fortunately, although the console shows the exception, this happens in a promise that doesn't affect the test result.

etpinard replied:
I don't think we should worry about making `Plotly.restyle` not output errors when used with mocked gd.

etpinard commented on test/jasmine/tests/plot_api_test.js:
```diff
@@ -786,4 +786,29 @@ describe('Test plot api', function() {
+
+            Plotly.plot(gd, data).then(function() {
+                Plotly.newPlot(gd, data, { height: 50 }).then(function() {
+                    expect(gd._fullLayout.height).toBe(50);
```
we should also check to `svg` attributes too to make sure that it updated properly.

etpinard commented:
@n-riesco looks like you're on the right path. 

> Replaced { autosize: 'initial' } with the flag gd._fullLayout._initialAutoSizeIsDone

Yep. Great idea. :+1: 

> { autosize: false } the values of width and height undefined in  gd.layout will be autosized only once.

Hmm. Is this the current behavior? Shouldn't `{ autosize: false }` lead to `layout.width: 700` and `layout.height: 400` as per their defaults?

> { autosize: true } only autosizes the values of width and height undefined in gd.layout

Great.

Unfortunately, the plotly.js auto-size has very poor test coverage; I'm a little concerned with the PR. @n-riesco could make sure that all code paths in [`plotAutoSize`](https://github.com/plotly/plotly.js/pull/577/files#diff-ad4f76ccd6044ed16514297078e13b84R783) are tested under `Plotly.plot` and `Plotly.relayout`.

n-riesco commented:
On 30/05/16 17:32,  Étienne Tétreault-Pinard wrote:
> @n-riesco <https://github.com/n-riesco> looks like you're on the right path.
>     { autosize: false } the values of width and height undefined in gd.layout will be autosized only once.
> 
> Hmm. Is this the current behavior? Shouldn't |{ autosize: false }| lead to |layout.width: 700| and |layout.height: 400| as per their defaults?

The current behaviour is that if layout.width or layout.height are undefined, then layout.autosize is set to 'initial'.
When layout.autosize is set to 'initial' (and assuming config.fillFrame is false), plotAutosize gets the layout width and height using window.getComputedStyle(gd). Only if this fails, then the default width and height values are used.

> Unfortunately, the plotly.js auto-size has very poor test coverage; I'm a little concerned with the PR. @n-riesco <https://github.com/n-riesco> could make sure that all code paths in |plotAutoSize| <https://github.com/plotly/plotly.js/pull/577/files#diff-ad4f76ccd6044ed16514297078e13b84R783> are tested under |Plotly.plot| and |Plotly.relayout|.

OK. I'll add more tests.

alexcjohnson commented:
@n-riesco this is great, that's definitely where the autosize code belongs.

My only question (which goes along with @etpinard 's concern about test coverage - there are tons of cases to consider) is whether there are cases we'd like `autosize` to coerce to `true` - ie if width or height is missing? For example if someone is using element size to determine plot size, but has not explicitly set `autosize: true` - then something outside the plot (window resize or some event on their page) causes the element size to change, I would expect a simple redraw or relayout to resize the plot too.

This then begs the question (that we probably don't need to delve into now, might be a rabbit hole, but something to keep in mind for later) of whether we actually gain anything with the initial / redraw distinction. Either I'm missing something or it used to work differently, but I don't see that `plotAutoSize` does anything time-consuming, so what's to stop us from *always* running it if there's one or more missing dimensions or if `autosize` is explicitly `true`? There will be places (`relayout`?) where we need to track whether anything actually changed to decide how much to do next, but we already do that anyway.

n-riesco commented:
Below are the tests I'm planning to add:
- [ ] autosize:false, fillFrame:true
- [ ] autosize:false, fillFrame:false and frameMargins
- [ ] autosize:false, fillFrame:false and no frameMargins
- [ ] autosize:true, fillFrame:true
- [ ] autosize:true, fillFrame:false and frameMargins
- [ ] autosize:true, fillFrame:false and no frameMargins

For each test, I'll check fullLayout width and height after plot and after relayout.

These tests should cover the main execution paths in plotAutoSize.

etpinard commented:
@n-riesco your plans sounds good!

n-riesco commented:
While writing the tests for this PR, I noticed that `config` accepts the attribute `autosizable`. In master, this attribute is only honoured by `plotPolar`. Its meaning in `plotPolar` is:
- if `config.autosizable: false` and `layout.autosize: false`, then no initial autosize is carried out (if needed, the default values of width and height are used);
- if `config.autosizable: true`, then an initial autosize is carried out (regardless of `layout.autosize`).

I've updated this PR so that `autosizable` is always honoured.

n-riesco commented:
The PR is ready for review again.

etpinard commented:
@n-riesco Great. :beers: 

I'll keep focusing on our `mapbox-gl` integration [project](https://github.com/plotly/plotly.js/compare/mapbox-wip) for the rest of the week. 

I'll review this PR next Monday.

etpinard commented on src/plots/plots.js:
```diff
@@ -752,6 +782,99 @@ plots.supplyLayoutGlobalDefaults = function(layoutIn, layoutOut) {
+            factor = 1 - 2 * frameMargins;
+
+        var gdBB;
+        try {
```
I think we should avoid using `try-catch` as much as possible - because they are very slow.

Here too, I'd vote for using `Lib.isPlotDiv`. But, I'm open to other (possible more strict) suggestions.

n-riesco replied:
How about?
```js
var gdBB = fullLayout._container && fullLayout._container.node ?
    fullLayout._container.node().getBoundingClientRect() : {
        width: fullLayout.width,
        height: fullLayout.height
    };
```

etpinard commented on src/plot_api/plot_config.js:
```diff
@@ -25,7 +25,7 @@ module.exports = {
     editable: false,
 
-    // plot will respect layout.autosize=true and infer its container size
+    // DO initial autosize
```
maybe expand on that a little bit?

e.g. _make plotting code go through auto-size routine regardless of layout.autosize_

etpinard commented on src/plots/plots.js:
```diff
@@ -455,7 +464,7 @@ plots.sendDataToCloud = function(gd) {
     var oldFullLayout = gd._fullLayout || {},
         newFullLayout = gd._fullLayout = {},
-        newLayout = gd.layout || {};
+        layout = gd.layout || {};
```
I would keep the `newLayout` nomenclature here.

etpinard commented on src/plots/plots.js:
```diff
@@ -468,7 +477,29 @@ plots.supplyDefaults = function(gd) {
+
+    if(oldFullLayout._initialAutoSizeIsDone) {
+        // coerce the updated layout while preserving width and height
+        var oldWidth = oldFullLayout.width,
```
Why do we need this?

Doesn't `plotAutoSize` [copy](https://github.com/plotly/plotly.js/blob/19a37357341e7e098ce2c20ee98f65965c2c5d66/src/plot_api/plot_api.js#L2558-L2559) the calculated `width` and `height to `layout.width` and `layout.height` respectively?

etpinard replied:
Oh I see that you deleted that block.

I guess one less place where we `gd.layout` is mutated is a good thing. :+1:

etpinard replied:
... but, it might be considered backward incompatible. I think we wait for [`v2.0.0`](https://github.com/plotly/plotly.js/issues/420).

n-riesco replied:
I've restored backwards-compatibility and set width and height in `gd.layout` when `layout.autosize` is false.

etpinard commented on src/plots/plots.js:
```diff
@@ -522,6 +553,7 @@ plots.supplyDefaults = function(gd) {
     relinkPrivateKeys(newFullLayout, oldFullLayout);
 
+    // TODO may return a promise
     plots.doAutoMargin(gd);
```
Making this `gd._promises.push(plots.doAutoMargin(gd));` should do the trick.

n-riesco replied:
I tried:
```js
if(!Array.isArray(gd._promises)) gd._promises = [];
gd._promises.push(plots.doAutoMargin(gd));
```
but it caused a bunch of jasmine tests to fail.

I think this issue with `doAutoMargin` deserves its own PR, because I've seen other places where a promise is pushed into `gd._promises` and subsequently wiped out by a call to `Plotly.plot` (for example in `Plotly.relayout` there are cases where a call to `Image.draw` pushes a promise into `gd._promises` that subsequently gets wiped out by a call to `Plotly.plot`).

etpinard replied:
Ha. Good to know. Thanks for the clarification. Yes, leaving this for a separate PR is fine.

etpinard commented on src/plots/plots.js:
```diff
@@ -743,7 +773,7 @@ plots.supplyLayoutGlobalDefaults = function(layoutIn, layoutOut) {
 
     // called in plotAutoSize otherwise
-    if(autosize !== 'initial') plots.sanitizeMargins(layoutOut);
+    if(layoutOut.width && layoutOut.height) plots.sanitizeMargins(layoutOut);
```
Is this still necessary?

Looks like `layoutOut.width` and `layoutOut.height` will now always be defined at this stage.

n-riesco replied:
Great catch! That should be `layoutIn.width` and `layoutIn.height`.

etpinard commented on test/jasmine/tests/config_test.js:
```diff
@@ -1,10 +1,140 @@
+                },
+                layoutHeight = Plots.layoutAttributes.height.dflt,
+                relayoutHeight = layoutHeight;
+            testAutosize(autosize, config, layoutHeight, relayoutHeight, done);
```
amazing test cases!

etpinard commented:
> While writing the tests for this PR, I noticed that config accepts the attribute autosizable. In master, this attribute is only honoured by plotPolar. Its meaning in plotPolar is:

Yep. I knew that. I was thinking about dropping the `autosizable` config argument all together. But maybe it will useful for some users. @n-riesco what do you think?

n-riesco commented:
@etpinard About `autosizable`, I think it should be kept in Plotly v1, and removed in Plotly v2. The reason why I think it should be removed is that I don't think `autosizable` should be a global setting (different plots should be able to define different values of `autosizable`).

For Plotly v2, I would suggest to combine `autosize` and `autosizable` into a single setting. How about something like this?

```js
{
    autosize: {
        valType: 'enumerated',
        role: 'info',
        values: ['never', 'initial', 'always'],
        description: [
            'Determines how a layout width or height',
            'that has been left undefined by the user',
            'is set.',

            'If *never*, the default values of width and height will be used.',
            'If *initial*, an undefined layout width or height',
            'will be initialized on the first call to plot.'
            'If *always*, an undefined layout width or height',
            'will be set on the first call to plot and subsequent calls to relayout.'
        ].join(' ')
    }
}
```

n-riesco commented:
PR ready for review again.

etpinard commented on src/plots/plots.js:
```diff
@@ -806,15 +806,11 @@ plots.plotAutoSize = function plotAutoSize(gd, layout, fullLayout) {
-        } catch(err) {
-            gdBB = {
+        var gdBB = fullLayout._container && fullLayout._container.node ?
+            fullLayout._container.node().getBoundingClientRect() : {
```
Good. :+1:

etpinard commented on src/plots/plots.js:
```diff
@@ -536,7 +536,7 @@ plots.supplyDefaults = function(gd) {
 
     // finally, fill in the pieces of layout that may need to look at data
-    plots.supplyLayoutModuleDefaults(layout, newFullLayout, newFullData);
+    plots.supplyLayoutModuleDefaults(newLayout, newFullLayout, newFullData);
```
Thanks !

etpinard commented:
Thanks @n-riesco . you're almost there. 

As mentioned previously, this PR was set uncover a lot of un-tested code. Commits 172d81dfad29a31ca06c745e08f9e602ac90628c and 862578ee48ab4b02a3b655728bc38bb708eca41d were no exceptions.

Would you mind a few test cases (one per commit is fine) for the two commits above? To lock their behavior until `v2.0.0`.

n-riesco commented:
I've added tests for the commits you suggested and this uncovered an issue with 862578e and another with `Lib.isPlotDiv`. Both are now fixed and tested.

The PR is ready for review again.

etpinard commented:
Great. I tested your branch and I couldn't find any edge cases that you might have missed. :tada: 

This for sure wins the PR of the week :trophy:

I'll make two minor comments on the line diff (nothing major) and then it'll be time to :dancer: .

etpinard commented on src/plots/layout_attributes.js:
```diff
@@ -47,11 +47,16 @@ module.exports = {
         role: 'info',
-        // TODO: better handling of 'initial'
-        values: [true, false, 'initial'],
+        values: [false, true],
```
we should change the `valType` to `'boolean'` and remove the `values` field.

etpinard commented on src/plots/plots.js:
```diff
@@ -287,17 +287,26 @@ plots.resize = function(gd) {
+            // nor should it be included in the undo queue
+            gd.autoplay = true;
+
+            Plotly.plot(gd).then(function() {
```
@n-riesco could you clarify why `Plots.resize` needs to go through `Plotly.plot` instead of `Plotly.relayout` now?

I think it has to do with `fullLayout._initialAutoSizeIsDone`, but I just want to make sure.

n-riesco replied:
Both `Plotly.plot(gd)` and `Plotly.relayout(gd, { autosize: true })` would have the same effect in terms of resizing, but `Plotly.relayout(gd, { autosize: true })` would have the side-effect of setting `layout.autosize` to true. But:
- if `layout.autosize` was `true`, there is no need to set it to `true` again;
- and if `layout.autosize` was `false`, both `layout.width` and `layout.height` would've already been initialised and setting `layout.autosize` to `true` would have no use.

etpinard replied:
Very clear. Thanks!

etpinard commented on test/jasmine/tests/plots_test.js:
```diff
@@ -326,7 +383,7 @@ describe('Test Plots', function() {
             gd = createGraphDiv();
 
-            Plotly.plot(gd, [{ x: [1, 2, 3], y: [2, 3, 4] }], {})
+            Plotly.plot(gd, [{ x: [1, 2, 3], y: [2, 3, 4] }], { autosize: true })
```
@n-riesco , @chriddyp  and I are testing the `mapbox` branch (which is rebased on top of the latest `master`) in the plot.ly workspace at the moment and we some issues.

Looks like this PR here is backward incompatible and the line :arrow_double_up: is a symptom. 

More clearly,

```js
Plotly.plot(gd, data);
```

should go through the `plotAutoSize` routine on the initial `Plotly.plot` call.

I believe changing the `autosize` attribute `dflt` to `true` would solve this. What are your thoughts on that?

n-riesco replied:
@etpinard An alternative that would be fully backwards-compatible is that `Plot.resize(gd)` deletes `gd.layout.width` and `gd.layout.height` and calls `Plotly.relayout(gd, { autosize: true });`. I'll prepare a PR so that you can check that it'd fix the issue in the `mapbox` branch.

n-riesco replied:
See #629

etpinard commented:
As per-slack conversation with @n-riesco , I'm reverting this PR and will spend more time testing in plot.ly's workspace before the next minor release - to make sure we handle all edge cases in a backward-compatible way.
